src/rt/array/utils.d

```diff
@@ -0,0 +1,94 @@
+
+private ulong accumulatePure(string file, int line, string funcname, string name, ulong size) nothrow pure
+{
+    static ulong impureBypass(string file, int line, string funcname, string name, ulong size) @nogc nothrow
+    {
+        import core.internal.traits : externDFunc;
+
+        alias accumulate = externDFunc!("rt.profilegc.accumulate", void function(string file, uint line, string funcname, string type, ulong sz) @nogc nothrow);
+        accumulate(file, line, funcname, name, size);
+        return size;
+    }
+
+    auto func = cast(ulong function(string file, int line, string funcname, string name, ulong size) @nogc nothrow pure)&impureBypass;
+    return func(file, line, funcname, name, size);
+}
+
+/**
+ * TraceGC wrapper around runtime hook `Hook`.
+ * Params:
+ *  T = Type of hook to report to accumulate
+ *  Hook = The hook to wrap
+ *  errorMessage = The error message incase `version != D_TypeInfo`
+ *  file = File that called `HookTraceImpl`
+ *  line = Line inside of `file` that called `HookTraceImpl`
+ *  funcname = Function that called `HookTraceImpl`
+ *  parameters = Parameters that will be used to call `Hook`
+ * Bugs:
+ *  This function template was ported from a much older runtime hook that bypassed safety,
+ *  purity, and throwabilty checks. To prevent breaking existing code, this function template
+ *  is temporarily declared `@trusted pure nothrow` until the implementation can be brought up to modern D expectations.
+*/
+auto HookTraceImpl(T, alias Hook, string errorMessage)(string file, int line, string funcname, Parameters!Hook parameters) @trusted pure nothrow
+{
+    version (D_TypeInfo)
+    {
+        pragma(inline, false);
+        string name = T.stringof;
+
+        // FIXME: use rt.tracegc.accumulator when it is accessable in the future.
+        version (tracegc)
+        {
+            import core.stdc.stdio;
+
+            printf("%sTrace file = '%.*s' line = %d function = '%.*s' type = %.*s\n",
+                Hook.stringof.ptr,
+                file.length, file.ptr,
+                line,
+                funcname.length, funcname.ptr,
+                name.length, name.ptr
+            );
+        }
+
+        ulong currentlyAllocated = gcStatsPure().allocatedInCurrentThread;
+
+        scope(exit)
+        {
+            ulong size = gcStatsPure().allocatedInCurrentThread - currentlyAllocated;
+            if (size > 0)
+                if (!accumulatePure(file, line, funcname, name, size)) {
+                    // This 'if' and 'assert' is needed to force the compiler to not remove the call to
+                    // `accumulatePure`. It really want to do that while optimizing as the function is
+                    // `pure` and it does not influence the result of this hook.
+
+                    // `accumulatePure` returns the value of `size`, which can never be zero due to the
+                    // previous 'if'. So this assert will never be triggered.
+                    assert(0);
```

Comment:
BTW, reading this I think there's may be a deeper issue whenever we have 'fake' pure code which internally calls impure code.

I guess the intent of such code is to not influence purity of the caller (i.e. completely optimizing away the caller would be fine), but the callee is not really pure, so it should not be removed on it's own. I wonder whether this can be expressed in any theoretically sound way, @tgehr any insights here?

It seems to me that the problem is that we lie to the compiler and simply tell it `this function is pure`. Maybe we should rather tell it `Ignore purity checks for this call, but the function is not pure`. Intuitively I'd say a
```d 
pure void foo()
{
    __traits(forcepure, bar());
}
```
Should work: if bar is not inlined, this would tell the compiler to keep the call, but ignore purity for purity calculation of the callee. With inlining, the compiler should be able to see the side effects anyway and should not optimize these away.

Reply:
I can see how your `assert` code has the same effect though: It marks the return value as used, as long as the call is not inlined. When it's inlined, the compiler may remove the assert, but it will see the impure parts of accumulatePure and keep these. This can probably also somehow solved like this: https://stackoverflow.com/questions/40122141/preventing-compiler-optimizations-while-benchmarking, but I'd prefer a compiler hook which should also work for void functions.

Reply:
I've been thinking about introducing `HookCallExp` or a bool in `CallExp` that ignore the purity and somehow allowed the code to be inline, which is disallowed right now `pragma(inline, false)`, whilst being able to be detected as the hook (`HookCompoundStatement`?) so the CTFE intercept code works.

Reply:
That sounds like a great idea.

Reply:
> BTW, reading this I think there's may be a deeper issue whenever we have 'fake' pure code which internally calls impure code.
> 
> I guess the intent of such code is to not influence purity of the caller (i.e. completely optimizing away the caller would be fine), but the callee is not really pure, so it should not be removed on it's own. I wonder whether this can be expressed in any theoretically sound way, @tgehr any insights here?
> ...

@jpf91 This was what `__mutable` functions were for in the original `__mutable` proposal. Inlining for such functions is not so easy if you want to preserve constraints on evaluation order precisely.